Local K8S pool - pods got panic

### Environment info
- NooBaa Version: **5.2.2-39b106a**
- Platform: ** OpenShift 4.2**

### Actual behavior
1. pods gets into CrashLoopBackOff
local-noobaa-noobaa-0                                     0/1     CrashLoopBackOff   8          22h
local-noobaa-noobaa-1                                     0/1     CrashLoopBackOff   8          22h
local-noobaa-noobaa-2                                     0/1     CrashLoopBackOff   8          22h

### Expected behavior
1. stable 

### Steps to reproduce
1.

### More information - Screenshots / Logs / Other output

```Oct-24 11:09:40.522 [Agent/16]    [L0] core.rpc.rpc_base_conn:: RPC CONN CLOSE ON ERROR n2n://5db0449828f3dd0022ffbc59(2zp7hox) [Error: N2N ICE CLOSED]
Oct-24 11:09:40.522 [Agent/16]    [L0] core.rpc.rpc_n2n:: _close undefined
Oct-24 11:09:40.522 [Agent/16]    [L0] core.rpc.ice:: TLS CLOSED: tcp://18.189.49.97:0=>tcp://10.131.0.99:60588
Oct-24 11:09:40.560 [Endpoint/80]   [LOG] CONSOLE:: read_rand_seed: reading 32 bytes from /dev/random ...
Oct-24 11:09:40.562 [Endpoint/80]    [L0] core.endpoint.endpoint:: waiting for ssl certificates
Oct-24 11:09:40.562 [Agent/16]    [L0] Agent.s3-agent-local-noobaa-noobaa-0-noobaa_storage-f045ed39:: got message from endpoint process: { code: 'WAITING_FOR_CERTS' }
Oct-24 11:09:40.562 [Endpoint/80]   [LOG] CONSOLE:: read_rand_seed: got 32 bytes from /dev/random, total 32 ...
Oct-24 11:09:40.563 [Endpoint/80]   [LOG] CONSOLE:: init_rand_seed: seeding with 32 bytes
OpenSSL 1.1.0j  20 Nov 2018 seeding randomness
Oct-24 11:09:40.563 [Endpoint/80]   [LOG] CONSOLE:: init_rand_seed: done
Oct-24 11:09:40.563 [Agent/16] [ERROR] CONSOLE:: PANIC: process uncaughtException TypeError: Cannot read property 'send' of null
    at Agent._handle_s3rver_messages (/root/node_modules/noobaa-core/src/agent/agent.js:790:51)
    at ChildProcess.emit (events.js:182:13)
    at ChildProcess.EventEmitter.emit (domain.js:442:20)
    at emit (internal/child_process.js:812:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)```

--- a/src/agent/agent.js
+++ b/src/agent/agent.js
@@ -755,7 +755,7 @@ class Agent {
                 // run node process, inherit stdio and connect ipc channel for communication.
                 const s3rver_args = ['--s3_agent', '--address', this.master_address];
                 dbg.log0('starting s3 server with command: node ./src/s3/s3rver_starter', s3rver_args.join(' '));
-                this.endpoint_info.s3rver_process = child_process.fork('./src/s3/s3rver_starter', s3rver_args, {
+                const proc = child_process.fork('./src/s3/s3rver_starter', s3rver_args, {
                         stdio: ['inherit', 'inherit', 'inherit', 'ipc']
                     })
                     .on('message', this._handle_s3rver_messages.bind(this))
@@ -765,15 +765,21 @@ class Agent {
                     })
                     .once('exit', (code, signal) => {
                         const RETRY_PERIOD = 30000;
-                        dbg.warn(`s3rver process exited with code=${code} signal=${signal}. retry in ${RETRY_PERIOD / 1000} seconds`);
-                        this.endpoint_info.s3rver_process = null;
-                        P.delay(RETRY_PERIOD)
-                            .then(() => {
-                                if (this.enabled) {
-                                    this._enable_service();
-                                }
-                            });
+                        dbg.warn(`s3rver process ${proc.pid} exited with code=${code} signal=${signal}`);
+                        if (!this.endpoint_info.s3rver_process ||
+                            this.endpoint_info.s3rver_process === proc) {
+                            dbg.warn(`active s3rver process exited and will retry in ${RETRY_PERIOD / 1000} seconds`);
+                            this.endpoint_info.s3rver_process = null;
+                            P.delay(RETRY_PERIOD)
+                                .then(() => {
+                                    if (this.enabled) {
+                                        this._enable_service();
+                                    }
+                                });
+                        }
                     });
+                this.endpoint_info.s3rver_process = proc;
+                dbg.log0(`started s3rver process with pid ${proc.pid}`);
             } catch (err) {
                 dbg.error('got error when spawning s3rver:', err);
             }
@@ -787,14 +793,18 @@ class Agent {
         dbg.log0(`got message from endpoint process:`, message);
         switch (message.code) {
             case 'WAITING_FOR_CERTS':
-                this.endpoint_info.s3rver_process.send({
-                    message: 'run_server',
-                    base_address: this.master_address,
-                    certs: this._ssl_certs,
-                    location_info: this.location_info,
-                    routing_hint: this.rpc.routing_hint,
-                    virtual_hosts: this.virtual_hosts
-                });
+                if (this.endpoint_info.s3rver_process) {
+                    this.endpoint_info.s3rver_process.send({
+                        message: 'run_server',
+                        base_address: this.master_address,
+                        certs: this._ssl_certs,
+                        location_info: this.location_info,
+                        routing_hint: this.rpc.routing_hint,
+                        virtual_hosts: this.virtual_hosts
+                    });
+                } else {
+                    dbg.warn('no active s3rver process. dropping message', message);
+                }
                 break;
             case 'STATS':
                 // we also get here the name of the bucket and access key, but ignore it for now
@@ -813,8 +823,12 @@ class Agent {
                     message: message.err_msg
                 };
                 // kill s3rver process on error
-                dbg.error('killing s3rver process', this.endpoint_info.s3rver_process.pid);
-                this.endpoint_info.s3rver_process.kill();
+                if (this.endpoint_info.s3rver_process && message.pid === this.endpoint_info.s3rver_process.pid) {
+                    dbg.error('killing s3rver process', this.endpoint_info.s3rver_process.pid);
+                    this.endpoint_info.s3rver_process.kill();
+                } else {
+                    dbg.warn('did not match a s3rver process to the relvant pid. got message:', message);
+                }
                 break;
             case 'STARTED_SUCCESSFULLY':
                 dbg.log0('S3 server started successfully. clearing srv_error');

--- a/src/endpoint/endpoint.js
+++ b/src/endpoint/endpoint.js
@@ -211,7 +211,8 @@ function handle_server_error(err) {
         process.send({
             code: 'SRV_ERROR',
             err_code: err.code,
-            err_msg: err.message
+            err_msg: err.message,
+            pid: process.pid,
         });
     } else {
         process.exit(1);

